Sources/Sentry/SentrySerialization.m

```diff
@@ -310,6 +292,153 @@ + (SentryLevel)levelFromData:(NSData *)eventEnvelopeItemData
+
+    SentryTraceContext *traceContext = envelope.header.traceContext;
+    if (traceContext != nil) {
+        [serializedData setValue:[traceContext serialize] forKey:@"trace"];
+    }
+
+    NSDate *sentAt = envelope.header.sentAt;
+    if (sentAt != nil) {
+        [serializedData setValue:sentry_toIso8601String(sentAt) forKey:@"sent_at"];
+    }
+    return serializedData;
+}
+
++ (NSData *)newlineData
+{
+    static NSData *newlineData = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{ newlineData = [@"\n" dataUsingEncoding:NSUTF8StringEncoding]; });
+    return newlineData;
+}
+
++ (BOOL)writeEnvelope:(SentryEnvelope *)envelope toPath:(NSString *)path
+{
+    // Make sure we don't leave partial envelopes
+    NSString *tempPath = [path stringByAppendingString:@".tmp"];
+
+    if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
+        NSError *deleteError = nil;
+        [[NSFileManager defaultManager] removeItemAtPath:tempPath error:&deleteError];
+
+        if (deleteError) {
+            SENTRY_LOG_ERROR(@"Could not make sure files do not exists.");
+            return NO;
+        }
+    }
+
+    NSOutputStream *outputStream = [NSOutputStream outputStreamToFileAtPath:tempPath append:NO];
+    if (!outputStream) {
+        SENTRY_LOG_ERROR(@"Failed to create output stream for path: %@", tempPath);
+        return NO;
+    }
+    [outputStream open];
+
+    NSMutableDictionary *serializedData = [self buildEnvelopeSerialized:envelope];
+    BOOL headerSuccess = [self writeObject:serializedData toStream:outputStream];
+    if (!headerSuccess) {
+        [self cleanupFailedWrite:outputStream tempPath:tempPath];
+        SENTRY_LOG_ERROR(@"Envelope header cannot be converted to JSON.");
+        return NO;
+    }
+
+    for (SentryEnvelopeItem *envelopeItem in envelope.items) {
+        if (![self writeNewlineToStream:outputStream tempPath:tempPath]) {
+            return NO;
+        }
+
+        NSDictionary *serializedItemHeaderData = [envelopeItem.header serialize];
+        BOOL itemSuccess = [self writeObject:serializedItemHeaderData toStream:outputStream];
+        if (!itemSuccess) {
+            [self cleanupFailedWrite:outputStream tempPath:tempPath];
+            SENTRY_LOG_ERROR(@"Envelope item header cannot be converted to JSON.");
+            return NO;
+        }
+
+        if (![self writeNewlineToStream:outputStream tempPath:tempPath]) {
+            return NO;
+        }
+        NSInteger writeResult = [outputStream write:envelopeItem.data.bytes
+                                          maxLength:envelopeItem.data.length];
+        if (writeResult < 0) {
+            [self cleanupFailedWrite:outputStream tempPath:tempPath];
+            SENTRY_LOG_ERROR(@"Failed to write item data to stream.");
+            return NO;
+        }
+    }
+
+    [outputStream close];
+
+    NSError *moveError;
+    [[NSFileManager defaultManager] moveItemAtPath:tempPath toPath:path error:&moveError];
```

Comment:
`m`: Please check if the `SentryFileManager` offers functionality that can be used here too.